File: today4/binary-search-tree-iterator.ts
/**
 *
 */
export { BSTIterator, TreeNode };

class BSTIterator {
  stack: TreeNode[] = [];

  constructor(root: TreeNode | null) {
    let pointer = root;
    while (pointer) {
      this.stack.push(pointer!);
      pointer = pointer!.left;
    }
  }

  next(): number {
    const next = this.stack.pop();
    if (next!.right) this.stack.push(next!.right);
    return next!.val;
  }

  hasNext(): boolean {
    return !!this.stack.length;
  }
}

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

File: returned/python/today14/binary_search_tree_iterator.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        self.stack = []
        def dfs(node: TreeNode):
            if not node:
                return
            dfs(node.right)
            self.stack.append(node.val)
            dfs(node.left)
        dfs(root)

    def next(self) -> int:
        return self.stack.pop()

    def hasNext(self) -> bool:
        return len(self.stack) > 0